### ISSUE TITLE
What type of PipeOp is composition?

### ISSUE THREAD

**Issue by RaphaelS1** [mlr-org/mlr3proba#35]
For implementing composition from `lp` to `distr` (see #31) we are building a compositor pipeline. We could argue that the compositor takes multiple learners, e.g. a `surv.rpart` and a `surv.kaplan` and therefore can inherit from `PipeOpEnsemble` but it may be simpler to only take one learner, in this example `surv.rpart` and just use `survfit` to estimate the baseline hazard.

The question is then, even if it only takes one input, can we still class composition as an ensemble method? If no then a new pipe op is needed, `PipeOpComposition`. 

Advantage for using `PipeOpEnsemble`:
* It takes as input prediction objects and outputs another prediction object, which is exactly what is required in this case
* Composition can be viewed as an ensemble method if we are using multiple learners to do so, which seems natural but possible over-complicated

Advantage for using `PipeOpComposition`:
* A generic composition PipeOp may allow for more flexibility down the line, however as each input/output type may be different, this is perhaps not in line with the current `mlr3pipelines` design in which the parent class appears to determine input/output numbers and type

I think perhaps starting with `PipeOpEnsemble` is the simplest way forward for now and we could consider abstracting to a generic class later on.

**Comment by RaphaelS1** [comment]
As an argument against `PipeOpEnsemble`, this currently assumes that the predict type is the same in and out. Of course this can easily be changed by adding another column, but the currently implemented ensemble methods include majority vote (for classification) and mean response (for regression). So predict type in = predict type out. Whereas in this case we are going from lp/crank in to lp/crank *and* distr out.

**Comment by fkiraly** [comment]
I think it should be a special, new type of node.
Reduction and composition should probably be separate.

Composition should be probably distinguished (as property) by component types, and whether it is component type homogenous or component type heterogeneous.

Ensembling is a kind of composition - usually the components are homogenous of type - but not all composition is ensembling, not even all type homogenous compositions are ensembling.

**Comment by RaphaelS1** [comment]
![Header_1154589_1 1-1023x1024](https://user-images.githubusercontent.com/25639974/68014426-7548ca80-fc87-11e9-9601-f54f4cface42.jpg)

**Comment by fkiraly** [comment]
well, you can't avoid to eat some cans of worms...

**Comment by fkiraly** [comment]
I also don't think having a new `PipeOpComposition` is substantially beyond the existing design, or the ensemble type, for that - from my vague viewpoint, it looks like it's just a minor extension/generalization of signature convention, interface, and properties.

**Comment by RaphaelS1** [comment]
Okay, so first to clarify terminology, I assume what you call a "node" is what pipelines calls a `PipeOp`. A `PipeOp` (pipe operator) is an operation in a graph, which I believe is the same as a node.

In theory I don't see why we can't have a new parent class `PipeOp` for composition `PipeOpCompositor` and reduction `PipeOpReductor` (reducer?), however I have three immediate questions from this:
1. As these would in theory be useful to all aspects of mlr3 should these parent classes not live in `mlr3pipelines` (as opposed to `mlr3proba`)
2. How can we best describe to a user the difference between composition and ensembling and why one doesn't inherit from the other?
3. Is it okay that the child classes of these may have different input/output types? (This is a more general question about pipelines)

**Comment by fkiraly** [comment]
>  I assume what you call a "node" is what pipelines calls a PipeOp.

Yes, sorry - I was mixing this a little bit with @ablaom 's learning network terminology.
In the mlr3design this would be `PipeOp`, but that's essentially a node in a graph of abstract model composition instructions, hence "node".

1: I would be strongly in favour that once design is consolidated, it moves to mlr3pipelines. Why: composition is a generic concept, not specific to probabilistic regression, and useful for vanilla tasks like classification etc too, though it's ultimately up to @berndbischl .
2. Formally, ensembling is a type of composition, which could be taken to imply an inheritance direction. I don't appreciate the full mlr3pipelines design sufficiently though to suggest this seriously.
3. In line with generic toolbox design principles, you could have sub-trees in the inheritance tree where the input/output types are the same. Similar how all learners inherit from the generic, but classifiers inherit from a classifier ancestor which in turn inherits fro the generic. Not sure about the best hierarchy though.

**Comment by RaphaelS1** [comment]
Okay, my suggestion is as follows, in order to make this simpler for Bernd and Michel.

* We implement `PipeOpCompositor`
* We implement `PipeOpDistrCompose` as a child-class of `PipeOpCompositor`

We demonstrate how `PipeOpEnsemble` can be a child-class of `PipeOpCompositor` as well (in abstract without actually changing the code). 

If @berndbischl & @mllg agree, then we create `PipeOpReductor` , move this and `PipeOpCompositor` to `mlr3pipeplines`, set ensemble to inherit from compositor, and keep the proba specific pipeops (e.g. distr compositor) in `mlr3proba`

**Comment by berndbischl** [comment]
I am happy to help. But I don't exactly understand what you want. Could you pls precisely explain what composition should do as a pipeop? And yeah that's a node in our graph.

**Comment by fkiraly** [comment]
An abstract composition node "eats" the parent nodes (as learners) and "produces" a learner at a child node, the "composite".

The learner type, i.e., which task it solves, or which output/predict type it has, can be different for parent nodes, or child node.

As a default, the child/composite will have the hyper-parameters of all the parents, plus the hyper-parameters of the composition node.

**Comment by berndbischl** [comment]
Sorry maybe I am dump I still don't get the first paragraph. Could you pls give 1 or 2 examples. 

 I know what functional composition is. Is this related?

**Comment by RaphaelS1** [comment]
A composition is just taking one learner and doing something outside of it's internal train/predict to transform it in some way. Some examples:

1. Survival Distribution

  * Predict a relative risk
  * Assume a proportional hazards model
  * Estimate baseline with Kaplan-Meier
  * Use composition to multiply the relative risk by kaplan-meier estimator and return a distribution

2. Ensembling

  * Predict response for two learners
  * Compose the two learners by aggregating the responses

**Comment by fkiraly** [comment]
Formally, you can see composition as an abstract construction process via a function

function compose
inputs: learner1 of type A1, learner2 of type A2, ..., learnerN of type AN
output: learner of type B

This can be done via a standalone function (e.g., wrapper) which is not so elegant, or as a constructor of a learner ComposedLearner.

Even more elegantly, you could consider the composition step being a node in a pipeline.

**Comment by fkiraly** [comment]
reduction is a special case of composition, where N=1, and you want to emphasize that input type is different from output type.

Usually the "input learner task" is more less difficult than the "output learner task"; since you solve the output task type by the input task type, but equally you can leverage a more complex learner to solve an easier task.

Examples are:
- thresholding in probabilistic classification, to get a deterministic classifier. This reduces to the more difficult task, but "canonically", and introduces the threshold as a parameter.
- binning/histogram strategies for probabilistic regression, then applying probabilistic classifiers to bin containedness as a label. This reduces probabilistic regression to the easier task of probabilistic classification.

**Comment by mllg** [comment]
@mb706 your 2 cents?

**Comment by mb706** [comment]
One thing to be aware of is that the `PipeOpLearner` produces `Prediction` objects, but does not expose the learned model (I have some ideas that this could be changed (#285) but these are incomplete). So composition that takes multiple `Prediction`s and produces one `Prediction` out of these works trivially, and if that does the job I would probably support having `composition` as a base class to `ensemble`.

If the actual learned models are necessary (e.g. because they have to be evaluated at points suggested by another model) things would probably be more difficult. Maybe they could be implemented using a different Graphs.

"reduction" sounds like what we are doing in ordinal regression.

**Comment by berndbischl** [comment]
sorry for late reply.
reading the latest comments from @fkiraly and @RaphaelS1 i now see it like this

- what you want and call "composition" seems to mean composition more in the sense of OO-programming then functional composition. where a class (here: node) is made up of more complex member vars. in your case you seem to specifically want to consider 1 or multiple learners inside of the PipeOp. Ensembles or "functionality wrappers" around single learners seem obvious examples. Thats fine and understandable.

- you then seem to propose abstract base classes for that, to concretely inherit from same. 

- if you want this, this should be discussed certainly not here, but in the pipelines repo. i would strongly argue not to introduce non-standard abstract structures here in a specific scenario package downstream. as what we are discussing is very general and not specific to "proba".

- we try to avoid such wrapper in mlr3pipelines as it is much harder to access data in them. eg. have a look at how we currently implement ensembles. no composition, no wrappers, a simple flat structure. i like that.

- i dont think we can avoid "learner composition" / wrapping completely, eg one case i discussed with @mb706 multiple times is target transformation. there you have to do something before you go into training (log-y), but also after prediction (exp-yhat). thats not so cool in a flat graph, where the node must come either before OR after - or we have to create 2 nodes, which looks not very elegant.

i am happy to discuss this further with you. but we need to do this in general, and we might need a call

**Comment by RaphaelS1** [comment]
Okay can we add this as an agenda point to the next mlr meeting or would you rather set-up a separate call? This is quite an important point for the next part of mlr3proba development and we can't release to CRAN without it

**Comment by berndbischl** [comment]
I am against discussing this during the normal calls. They are super full anyway. You need dedicated persons and time for this. Sometimes even a call is bad instead of meeting in person, but I think we cannot easily change this

**Comment by RaphaelS1** [comment]
Sure that makes sense. Who is in charge of pipelines then?

**Comment by berndbischl** [comment]
mainly @mb706, but for design you can include me and @mllg

**Comment by fkiraly** [comment]
@berndbischl, may I ask you to point me to:
- code
- design documents (if existing)
that explain your general pipeline design, and why, in your current design, you cannot handle target transformations easily?

I'm aware that sklearn and the prevalent designs in general have problems with that, but I'd personally put that down to mistakes in designing the pipeline specification formalism.

For example, in the "right" formalism (that I have in mind and which I boldly attach positive adjectives to), target transformation is just one node *after* the predictive learner it applies to.

The higher level question here is (which I'd also like to understand better):
what are *advantages* of making operations to the different arguments of the learners explicit in the graph structure, rather than operations on the level of learners? As long as the latter they are correctly and efficiently translated (or "compiled" if you so will) to the former?

**Comment by fkiraly** [comment]
For an interim solution: can we not just write constrution wrappers?

Their final state in the pipeline and oop paradigm remaining subject to discussion.

**Comment by berndbischl** [comment]
can we pls dicuss this with enough time in a call? i think this gets too complex for chatting in this issue here

**Comment by berndbischl** [comment]
@fkiraly have you had a look at 
a) the new overview talk on mlr3pipelines as pdf - shows the concept
b) the technical r docs

of course both a) and b) really do not talk about "what we currently cannot do". see request for call above

**Comment by fkiraly** [comment]
link to "new overview talk"?
re. call, see slack.

**Comment by RaphaelS1** [comment]
I agree with Bernd and think we should postpone any further discussion until a call. Until then I will prepare some clear documentation for what I am proposing, and how this relates to the current designs and infrastructure (of which I am increasingly familiar with)

**Comment by berndbischl** [comment]
> link to "new overview talk"?

pls read

1. this, all 3 slidesets
https://github.com/mlr-org/mlr-outreach/tree/master/2019_odsc/slides

2. then this. the text might not be perfect
https://mlr3pipelines.mlr-org.com/articles/introduction.html

3. then the technical docs of pipelines, especially the base classes

pls do this before the meeting

**Comment by fkiraly** [comment]
I've read the docs, but I was looking without success for specific information on how the most common composition operation - grid search CV - could map onto/into the `PipeOp` formalism. I think this is a crucial point for me to understand.

More precisely, is there a way to do CV *within* a network via `PipeOp`?

As far as I can see, there are only two ways to do this, which aren't strictly speaking `PipeOp`s:
- use an `AutoTuner` as a network component
- tune outside, i.e., wrap a graph in `GraphLearner` then `AutoTuner`

Thus, in the current design, general composition or tuning is *not meant to be* a `PipeOp` - unless you can express it as sequential transformation on train/predict.

This (restrictive?) design has as negative side effects:
- you can't deal with target transformation, because it acts *before* and *after* the same node, depending on learning phase - rather than on the same side in both phases as the design assumes.
- to deal with bagging (or other ensembling), you have to split the strategy up in multiple components and patch them back together, i.e., `subsample` and `greplicate` and `classifavg`. No single convenient wrapper is available, because you hate wrappers.

Is this correct?

**Comment by mb706** [comment]
> More precisely, is there a way to do CV within a network via PipeOp?

No, with the exceptions you have listed yourself.

> composition or tuning is not meant to be a PipeOp - unless you can express it as sequential transformation on train/predict.

Correct.

We have tried to keep the functionality of the `Graph` as simple as possible--PipeOps produce something that they then pass on along edges to other PipeOps; this happens exactly once per PipeOp and per train or predict cycle, and without PipeOps needing to access information from the Graph itself, or modifying the Graph.

Maybe it would be possible without much pain to get bagging to work from the Graph itself by introducing some kind of do-the-following-multiple-times PipeOp, and another PipeOp that then collects the multiple outputs; this would require some adjustments to the Graph similar to PipeOp(Un)Branch.

I dont see how we should implement more complicated compositon / target transformation in PipeOp form -- the problem here is that if we have a target transform, followed by *something*, followed by a PipeOpLearner, followed by *something else* (possibly an ensemble), it would not be obvious when to do the inverse target transform on the prediction generated by the Learner. mlrCPO did not have this issue because it did not allow ensembles (and forbade target CPOs inside concurrent parts of the preproc graphs). If you have ideas here please share!

**Comment by fkiraly** [comment]
>If you have ideas here please share!

I do - let's perhaps discuss in the scheduled call.
I can also write up sth until then with @RaphaelS1 - to send, to discuss, or to later post in an issue if there's one for this.

**Comment by fkiraly** [comment]
I notice we haven't recorded the outcome of our discussion, let me try:
- generic composition is not in general a `PipeOp`
- attaching some `PipeOp`s is composition, and so is a combined operation attaching specific `PipeOps` (e.g., ensembling)
- for the moment, we should avoid extending `PipeOp`s, but use `PipeOp`s, or `PipeOp` sugar abstractions wherever they can be used. Otherwise, compositors should follow `AutoTuner` blueprint.
- we plan to review model composition syntax and architectural formalisms on mid-term

(?)
